### PR TITLE
DEX-21 and DEX-22 functionality

### PR DESCRIPTION
**Pull request by naknomum**
creates `/api/v1/submissions/tus` and `/api/v1/submissions/tus/collect` to support [tus](https://tus.io/) / [uppy](https://uppy.io/) on frontend.

note: this requires **redis** to be installed on houston server.

### CODE REVIEW

File: app/extensions/submission/flask_tus_cont.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+# Find the stack on which we want to store the database connection.
+# Starting with Flask 0.9, the _app_ctx_stack is the correct one,
+# before that we need to use the _request_ctx_stack.
+try:
+    from flask import _app_ctx_stack as stack
+except ImportError:
+    from flask import _request_ctx_stack as stack
+
+
+class TusManager(object):
```
Comment (bluemellophone) [inline app/extensions/submission/flask_tus_cont.py:17]: Let's move this to a new extension as `app/extensions/tus/`

File: app/extensions/submission/__init__.py [hunk 2]
```diff
@@ -238,6 +238,11 @@ def ensure_repository(self, submission, remote=True):
         with open(submission_metadata_path, 'w') as submission_metadata_file:
             json.dump(submission_metadata, submission_metadata_file)
 
+        # gitignore_path = os.path.join(submission_path, '.gitignore')
```
Comment (bluemellophone) [inline app/extensions/submission/__init__.py:241]: If this code is not needed, let's delete it

File: tests/modules/assets/resources/test_assets.py [hunk 3]
```diff
@@ -146,12 +146,16 @@ def test_read_all_assets(
         assert admin_response.status_code == 200
         assert admin_response.content_type == 'application/json'
         assert len(admin_response.json) == 2
-        # @todo, is the order received deterministic
+        # update: it seems this order received _is not_ deterministic
```
Comment (bluemellophone) [inline tests/modules/assets/resources/test_assets.py:149]: Let's add `Asset.query.order_by(Asset.guid).offset(args['offset']).limit(args['limit'])`

File: app/extensions/submission/__init__.py [hunk 4]
```diff
@@ -302,3 +307,41 @@ def init_app(app, **kwargs):
     API extension initialization point.
     """
     SubmissionManager(app)
+    init_tus(app)
```
Comment (bluemellophone) [inline app/extensions/submission/__init__.py:310]: Likewise, let's move this code to a new extension just for TUS

File: app/modules/submissions/models.py [hunk 5]
```diff
@@ -515,3 +516,10 @@ def delete(self):
         db.session.refresh(self)
         with db.session.begin():
             db.session.delete(self)
+
+    @classmethod
+    def tus_upload_dir(cls, guid=None):
+        d = os.path.join('_db', 'uploads')
```
Comment (bluemellophone) [inline app/modules/submissions/models.py:522]: We need a new TUS config class defined in `config.py` for things like this to specify the upload path on disk

File: app/modules/submissions/resources.py [hunk 6]
```diff
@@ -244,3 +245,55 @@ def delete(self, submission):
...
+            # We have checked the submission manager and cannot find this submission, raise 404 manually
+            raise werkzeug.exceptions.NotFound
+
+        repo, project = submission.ensure_repository()
+        updir = Submission.tus_upload_dir(submission.guid)
+        submission_abspath = submission.get_absolute_path()
+        submission_path = os.path.join(submission_abspath, '_submission')
+        ct = 0
+        for root, dirs, files in os.walk(updir):
+            ct = len(files)
```
Comment (bluemellophone) [inline app/modules/submissions/resources.py:283]: What is `ct` and is this meant to be a `+=`?

File: app/static/uppy-test.html [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
```
Comment (bluemellophone) [inline app/static/uppy-test.html:1]: This file should be moved to a template using the `.jinja2` format (see `app/templates/home.jinja2` for an example) and an accompanying route added to `app/modules/frontend/views_backend.py` to support this Houston-only test interface

File: app/modules/submissions/resources.py (hunk 6 above)
Comment (naknomum) [reply app/modules/submissions/resources.py:285]: nope just meant to be "are there any files (and if so how many)"

File: app/extensions/submission/__init__.py (hunk 2 above)
Comment (naknomum) [reply app/extensions/submission/__init__.py:241]: yeah will remove for now.  was hoping to get a `.gitignore` in there to keep some stuff out of repo.  but it didnt work.  meh.  later.

File: app/extensions/tus/flask_tus_cont.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+            os.path.lexists(os.path.join(self.upload_folder, metadata.get('filename')))
+            and self.file_overwrite is False
+        ):
+            response.status_code = 409
+            return response
+
+        file_size = int(request.headers.get('Upload-Length', '0'))
+        resource_id = self.create_resource_id()
+
+        p = self.redis_connection.pipeline()
```
Comment (hwindsor) [inline app/extensions/tus/flask_tus_cont.py:204]: Could p have a better name? I've never liked single char variable names

Comment (naknomum) [reply app/extensions/tus/flask_tus_cont.py:204]: agree with you conceptually here.... backstory on this code is it is a [clone of a library](https://pypi.org/project/Flask-Tus-Cont/) we needed to make a couple tweaks to.  so i am tempted to keep it the same.  (why?  i dunno... i doubt we are ever going to look to patch changes back in...).  counter-argument is that _linting_ has already moved things around in their code anyway!  haha. maybe i should drop a README in this dir explaining this history?

Comment (hwindsor) [reply app/extensions/tus/flask_tus_cont.py:204]: Aha, if it's inherited and tweaked then yes, a README pointing to the source and documenting any changes sounds like a much better idea.

Comment (naknomum) [reply app/extensions/tus/flask_tus_cont.py:204]: i have added this readme

**Review by colinwkingen — APPROVED** [review]
LGTM. Adds the README update and the further tweaks for sometime-failing test is addressed in a different PR.

File: app/modules/submissions/models.py (hunk 5 above)
Comment (naknomum) [reply app/modules/submissions/models.py]: @bluemellophone  -  just fyi, i have made this change as part of my current feature branch for sightings.  look for it in a develop branch near you soon!   thanks for the suggestion

https://github.com/WildMeOrg/houston/blob/sighting_passthrough_basic/app/extensions/tus/__init__.py#L64

**Pull request merged**